air/src/execution_step/instructions/canon.rs

```diff
@@ -40,17 +42,56 @@ impl<'i> super::ExecutableInstruction<'i> for ast::Canon<'i> {
         log_instruction!(call, exec_ctx, trace_ctx);
         let canon_result = trace_to_exec_err!(trace_ctx.meet_canon_start(), self)?;
 
+        let epilog: &dyn for<'ctx> Fn(
+            StreamWithSerializedView,
+            &mut ExecutionCtx<'_>,
+            &mut TraceHandler,
+        ) -> ExecutionResult<()> = &|stream_with_positions: StreamWithSerializedView,
+                                     exec_ctx: &mut ExecutionCtx<'_>,
+                                     trace_ctx: &mut TraceHandler|
+         -> ExecutionResult<()> {
+            let StreamWithSerializedView {
+                canon_stream,
+                canon_result_cid,
+            } = stream_with_positions;
+
+            exec_ctx.scalars.set_canon_value(
+                self.canon_stream.name,
+                CanonStreamWithProvenance::new(canon_stream, canon_result_cid.clone()),
+            )?;
+
+            trace_ctx.meet_canon_end(CanonResult::new(canon_result_cid));
+            Ok(())
+        };
+
         match canon_result {
             MergerCanonResult::CanonResult(canon_result_cid) => {
-                handle_seen_canon(self, canon_result_cid, exec_ctx, trace_ctx)
+                handle_seen_canon(epilog, canon_result_cid, exec_ctx, trace_ctx)
+            }
+            MergerCanonResult::Empty => {
+                // This closure gets a stream from a context or returns a default empty stream,
+                // it is crucial for deterministic behaviour, for more info see
+                // github.com/fluencelabs/aquavm/issues/346
+                let get_stream_or_default: &dyn for<'ctx> Fn(&'ctx mut ExecutionCtx<'_>) -> Cow<'ctx, Stream> =
```

Comment:
For better code readability, I suggest to leaving this as a separate function that will return a closure.

Reply:
Pardon my ignorance, but I am somewhat skeptical that a closure that returns Fn trait object or a Box improves readability :)
Will try it though

Reply:
Done